state-sync/storage-service/server/src/moderator.rs

```diff
@@ -0,0 +1,395 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{error::Error, logging::LogEntry, metrics, LogSchema};
+use aptos_config::{
+    config::StorageServiceConfig,
+    network_id::{NetworkId, PeerNetworkId},
+};
+use aptos_infallible::RwLock;
+use aptos_logger::warn;
+use aptos_network::application::storage::PeersAndMetadata;
+use aptos_storage_service_types::{
+    requests::StorageServiceRequest, responses::StorageServerSummary,
+};
+use aptos_time_service::{TimeService, TimeServiceTrait};
+use std::{
+    collections::HashMap,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+/// A simple struct that tracks the state of an unhealthy peer
+#[derive(Clone, Debug)]
+pub struct UnhealthyPeerState {
+    ignore_start_time: Option<Instant>, // The time when we first started ignoring the peer
+    invalid_request_count: u64,         // The total number of invalid requests from the peer
+    max_invalid_requests: u64, // The max number of invalid requests before ignoring the peer
+    min_time_to_ignore_secs: u64, // The min time (secs) to ignore the peer (doubles each round)
+    time_service: TimeService, // The time service
+}
+
+impl UnhealthyPeerState {
+    pub fn new(
+        max_invalid_requests: u64,
+        min_time_to_ignore_secs: u64,
+        time_service: TimeService,
+    ) -> Self {
+        Self {
+            ignore_start_time: None,
+            invalid_request_count: 0,
+            max_invalid_requests,
+            min_time_to_ignore_secs,
+            time_service,
+        }
+    }
+
+    /// Increments the invalid request count for the peer and marks
+    /// the peer to be ignored if it has sent too many invalid requests.
+    /// Note: we only ignore peers on the public network.
+    pub fn increment_invalid_request_count(&mut self, peer_network_id: &PeerNetworkId) {
+        // Increment the invalid request count
+        self.invalid_request_count += 1;
+
+        // If the peer is a PFN and has sent too many invalid requests, start ignoring it
+        if self.ignore_start_time.is_none()
```

Comment:
Yeah, we do this already when the `validate_request` method fails and returns an error to the storage service. Annoyingly, we have to sample though and can't do it for every request because it produces really spammy logs. 😢

Reply:
nit, I'd put `self.invalid_request_count >= self.max_invalid_requests` first just in case it compiles and checks that first as the cheapest check